### DISCUSSION TITLE
quadlet:  environment variable substitution works differently for `podman.addf("-w=%s", workdir)` and `podman.add("--workdir", workdir)`

### DISCUSSION THREAD

**Discussion by eriksjolund** [containers/podman#24105] [General]
Environment variable substitution in systemd user service works differently
 for `podman.addf("-w=%s", workdir)` and `podman.add("--workdir", workdir)`

## test environment variable substitution in systemd user service

I perfomed these steps:

1. `sudo useradd test`
2. `sudo machinectl shell --uid=test`
3. `mkdir -p ~/.config/systemd/user`
4. Create the file _/var/home/test/.config/systemd/user/test.service_ containing
   ```
   [Service]
   Type=oneshot
   ExecStart=/usr/bin/echo foobar -w=$HOME -w=${HOME} --workdir $HOME --workdir ${HOME}
   ```
5. `systemctl --user daemon-reload`
6. `systemctl --user start test.service`
7. Check the journal
   ```
   journalctl --user --no-pager -xeu test.service | grep foobar
   ```
   The command prints the following output
   ```
   Sep 29 16:28:08 localhost.localdomain echo[17465]: foobar -w=$HOME -w=/var/home/test --workdir /var/home/test --workdir /var/home/test
   ```

__summary of the result:__

| input | result after substitution |
|               ---- |    --  |
|   -w=$HOME     | -w=$HOME   |
|  -w=${HOME}    | -w=/var/home/test |
|   --workdir $HOME     | --workdir /var/home/test |
|   --workdir ${HOME}     | --workdir /var/home/test |

## Discussion

Variable substitution thus behaves differently depending on whether _quadlet.go_ implements `WorkingDir=` with

```
podman.addf("-w=%s", workdir)
```

or

```
podman.add("--workdir", workdir)
 ```

Currently the first alternative is used

https://github.com/containers/podman/blob/e1496c992acd54f75a54635f7c04915b64e906ad/pkg/systemd/quadlet/quadlet.go#L786
(a GitHub permalink as of 29 September 2024)

Because of that, the following example

```
WorkingDir=$HOME
```

in

* https://github.com/containers/podman/blob/e1496c992acd54f75a54635f7c04915b64e906ad/docs/source/markdown/podman-systemd.unit.5.md?plain=1#L336

does not work.

**Comment by Luap99** [comment]
This looks interesting, I guess it would make the most sense if quadlet behaves consistently for all options as this would effect all options not just workdir.

I would say `podman.add("--workdir", workdir)` makes more sense as this allows the "normal" `$var` syntax to work as most would expect.

cc @ygalblum

**Reply by ygalblum** [reply]
@eriksjolund Thanks for the analysis.
@Luap99, yes, I agree. Quadlet should be consistent. I'll take a look
